crypto/aes_gcm.go

```diff
@@ -0,0 +1,161 @@
+package crypto
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/json"
+	"fmt"
+	"hash"
+
+	"github.com/bndw/pick/errors"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+type AESGCMClient struct {
+	config   Config
+	settings AESGCMSettings
+}
+
+type AESGCMSettings struct {
+	KeyLen           int    `json:"keylen,omitempty" toml:"keylen"`
+	Pbkdf2Hash       string `json:"pbkdf2hash,omitempty" toml:"pbkdf2hash"`
+	Pbkdf2Iterations int    `json:"pbkdf2iterations,omitempty" toml:"pbkdf2iterations"`
+	Pbkdf2SaltLen    int    `json:"pbkdf2saltlen,omitempty" toml:"pbkdf2saltlen"`
+}
+
+type AESGCMStore struct {
+	Salt       []byte `json:"salt"`
+	Nonce      []byte `json:"nonce"`
+	Ciphertext []byte `json:"ciphertext"`
+}
+
+func DefaultAESGCMSettings() *AESGCMSettings {
+	return &AESGCMSettings{
+		KeyLen:           32,
+		Pbkdf2Hash:       "sha512",
+		Pbkdf2Iterations: 100000,
+		Pbkdf2SaltLen:    16,
+	}
+}
+
+func NewAESGCMClient(config Config, settings AESGCMSettings) (*AESGCMClient, error) {
+	return &AESGCMClient{
+		config:   config,
+		settings: settings,
+	}, nil
+}
+
+func (c *AESGCMClient) keyLen() int {
+	keyLen := c.settings.KeyLen
+	switch keyLen {
+	default:
+		if keyLen != 0 {
+			fmt.Println("Invalid keyLen, using default")
+		}
+		return 32
+	case 16:
+	case 24:
+	case 32:
+	}
+	return keyLen
+}
+
+func (c *AESGCMClient) pbkdf2HashFunc() func() hash.Hash {
+	pbkdf2Hash := c.settings.Pbkdf2Hash
+	switch pbkdf2Hash {
+	default:
+		if pbkdf2Hash != "" {
+			fmt.Println("Invalid PBKDF2 Hash, using default")
+		}
+		fallthrough
+	case "sha512":
+		return sha512.New
+	}
+}
+
+func (c *AESGCMClient) pbkdf2SaltLen() int {
+	return c.settings.Pbkdf2SaltLen
+}
+
+func (c *AESGCMClient) deriveKeyWithSalt(password []byte, salt []byte) []byte {
+	keyLen := c.keyLen()
+	pbkdf2Iterations := c.settings.Pbkdf2Iterations
+	return pbkdf2.Key([]byte(password), salt, pbkdf2Iterations, keyLen, c.pbkdf2HashFunc())
```

Comment:
`password` is already a `[]byte`

I'd either use all locals or all functions when calling `pbkdf2.Key`, but be consistent.